text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,203 @@
+[#1495](https://github.com/open-telemetry/opentelemetry-specification/pull/1495)
+renamed `cloud.zone` to `cloud.availability_zone`.  In a Schema world, this
+change could have been defined with the following schema rule:
+
+```yaml
+  next.version:
+    resources:
+      changes:
+        - rename_attributes:
+            attribute_map:
+              cloud.zone: cloud.availability_zone
+  previous.version:
+```
+
+If an exporter is defined expecting to find `cloud.zone` on `previous_version`,
+then when the SDK is bumped to `next.version` it will no longer be able to see
+`cloud.zone`.
+
+Specifically, we assume the following components could have independent
+versioning + release cycles:
+
+```
++-------------------+     +-----+     +----------+
+| Resource Detector | <-> | SDK | <-> | Exporter |
++-------------------+     +-----+     +----------+
+```
+
+## Explanation
+
+We propose a new set of SDK features:
+
+- A mechanism to "migrate" a resource up/down compatible schema changes.
+- A mechanism to determine if two schema URLs allow compatible changes.
+
+### Consuming resources at specific schema version
+
+The first is done by providing a new method against `Resource` in the SDK that
+will transform the attributes as necessary to match a given schema version.
+
+e.g. In Java you might have the following for resource:
+
+```java
+public abstract class Resource {
+    ... existing methods ...
+    /** The current schema version. */
+    public String getSchemaUrl();
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema url.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(String schemaUrl);
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema file.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(File schemaFile);
+}
+```
+
+This would allow a consumer of resources to ensure a stable view of resource
+labels, e.g.
+
+```java
+public static String MY_SCHEMA_VERSION="https://opentelemetry.io/schemas/1.1.0";
+public void myMetricMethod(MetricData metric) {
+    useStableResource(metric.getResource().convertToSchema(MY_SCHEMA_VERSION));
+}
+```
+
+### Understanding schema compatibility
+
+The second will be done via a new (public) SDK method that allows understanding
+whether or not two schemas are compatible, and which is the "newer" version.
+
+e.g. in Java you would have:
+
+```java
+public abstract class Schemas {
+    public static boolean isCompatible(String oldUrl, String newUrl);
+    public static boolean isNewer(String testUrl, String baseUrl);
+}
+```
+
+This could be used in the `Resource.merge` method to determine if resources
+can be converted to a compatible schema.  Specfically resource can be merged
+IFF:
+
+- The schemas are compatible
```

Comment:
Same question, is this something that we need to worry? Should we assume that converting to the "newest" is always possible?